### PR TITLE
ES6-Shim: `Map.get()`, `Array.find()` and others may return `undefined`

### PR DESCRIPTION
**Pull request by c7hm4r**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes:
  - [Array.prototype.find()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/find#Return_value)
  - JSDoc change for [Array.prototype.findIndex()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/findIndex#Return_value)
  - [Map.prototype.get()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map/get#Return_value)
  - [WeakMap.prototype.get()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/WeakMap/get#Return_value)
- [ ] Increase the version number in the header if appropriate.
- [ ] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dslint/dt.json" }`.

This pull request
- Changes the definitions of `Array<T>.prototype.find(…)`, `Map<K, T>.get(…)` and `WeakMap<K, T>.get(…)` to return `T | undefined`
- Does a fix in the JSDoc for `Array<T>.prototype.findIndex(…)`
- Enables `strictNullChecks` in `tsconfig.json`
- Does several fixes in the tests

### CODE REVIEW

File: types/es6-shim/package.json [hunk 1]
```diff
@@ -0,0 +1,5 @@
+{
+  "dependencies": {
+    "es6-shim": "0.34"
```
Comment (mhegazy) [inline types/es6-shim/package.json]: why is this here?

Comment (c7hm4r) [inline types/es6-shim/package.json]: I followed the check list: "Test the change in your own code. (Compile and run.)"
To run the tests, I needed to install `es6-shim`, that is the reason for the `package.json`. Shall it be removed?

Comment (mhegazy) [inline types/es6-shim/package.json]: A dependency should only be added if you are including types from this package. es6-shim has no types.

Comment (c7hm4r) [inline types/es6-shim/package.json]: Maybe I do not understand correctly, but I think `es6-shim` actually declares or extends types (if necessary) in the global scope (you do not have to explicitly import them). For example, `Array.prototype.values()` as defined in the ES6 specification p. 424 is not provided in node v7.9.0. When I run the tests without `require("es6-shim")` at the top of es6-shim-tests.js in node v7.9.0 it outputs this error:
```
.../DefinitelyTyped/types/es6-shim/es6-shim-tests.js:132
iterableIteratorOfPoint = arrayOfPoint.values();
                                       ^

TypeError: arrayOfPoint.values is not a function
...
```

Compiling the tests with tsc without having `require("es6-shim")` at the top gives:
```
$ node --max-old-space-size=8192 ~/.npm-packages/bin/tsc es6-shim-tests.ts --typeRoots .. | grep es6-shim
es6-shim-tests.ts(1,15): error TS2451: Cannot redeclare block-scoped variable 'require'.
es6-shim-tests.ts(20,5): error TS2451: Cannot redeclare block-scoped variable 'key'.
es6-shim-tests.ts(185,1): error TS2322: Type 'Promise<void>' is not assignable to type 'Promise<Point>'.
es6-shim-tests.ts(191,66): error TS2345: Argument of type 'PromiseLike<Point>' is not assignable to parameter of type 'Point | Thenable<P
oint>'.
es6-shim-tests.ts(196,37): error TS2345: Argument of type 'IterableShim<Point>' is not assignable to parameter of type '(Point | Thenable
<Point>)[]'.
es6-shim-tests.ts(198,31): error TS2345: Argument of type 'IterableShim<Point>' is not assignable to parameter of type '(Point | Thenable
<Point>)[]'.
es6-shim-tests.ts(199,1): error TS2322: Type 'Promise<{}>' is not assignable to type 'Promise<void>'.
es6-shim-tests.ts(202,34): error TS2345: Argument of type 'PromiseLike<Point>' is not assignable to parameter of type 'Point | Thenable<P
oint>'.
es6-shim-tests.ts(206,40): error TS2345: Argument of type 'Error' is not assignable to parameter of type 'Point'.
```
And of cause when you `require("es6-shim")`, you need to have it installed.

But that does not mean the `package.json` must be committed.

**Comment by aozgaa** [comment]
Pinging @rbuckton

File: types/es6-shim/index.d.ts [hunk 2]
```diff
@@ -594,16 +593,16 @@ interface SetConstructor {
 
 declare var Set: SetConstructor;
 
-interface WeakMap<K extends object, V> {
+interface WeakMap<K extends Object, V> {
```
Comment (rbuckton) [inline types/es6-shim/index.d.ts]: This should be left as `object`. The `object` type was introduced in TypeScript 2.2 and has a different meaning than `Object`. We can use TypeScript 2.2 features in this file as it has been marked with a `// TypeScript Version: 2.2` comment at the top of the file.

File: types/es6-shim/index.d.ts [hunk 3]
```diff
@@ -594,16 +593,16 @@ interface SetConstructor {
...
-    get(key: K): V;
+    get(key: K): V | undefined;
     has(key: K): boolean;
     set(key: K, value?: V): WeakMap<K, V>;
 }
 
 interface WeakMapConstructor {
-    new <K extends object, V>(): WeakMap<K, V>;
-    new <K extends object, V>(iterable: IterableShim<[K, V]>): WeakMap<K, V>;
+    new <K extends Object, V>(): WeakMap<K, V>;
```
Comment (rbuckton) [inline types/es6-shim/index.d.ts]: As above, this should remain `object`.

**Review by rbuckton — CHANGES_REQUESTED** [review]

File: types/es6-shim/index.d.ts [hunk 4]
```diff
@@ -556,7 +555,7 @@ interface Map<K, V> {
     clear(): void;
     delete(key: K): boolean;
     forEach(callbackfn: (value: V, index: K, map: Map<K, V>) => void, thisArg?: any): void;
-    get(key: K): V;
+    get(key: K): V | undefined;
     has(key: K): boolean;
     set(key: K, value?: V): Map<K, V>;
```
Comment (rbuckton) [inline types/es6-shim/index.d.ts]: Either `set` should not have an optional `value`, or the callback to `forEach` should type value as `V | undefined`. I would err on the side of the lib definition in TypeScript and remove the optionality of `value` in `set`.

File: types/es6-shim/index.d.ts [hunk 5]
```diff
@@ -596,7 +595,7 @@ declare var Set: SetConstructor;
 
 interface WeakMap<K extends object, V> {
     delete(key: K): boolean;
-    get(key: K): V;
+    get(key: K): V | undefined;
     has(key: K): boolean;
     set(key: K, value?: V): WeakMap<K, V>;
```
Comment (rbuckton) [inline types/es6-shim/index.d.ts]: Either `set` should not have an optional `value`, or the callback to `forEach` should type value as `V | undefined`. I would err on the side of the lib definition in TypeScript and remove the optionality of `value` in `set`.

**Review by rbuckton — CHANGES_REQUESTED** [review]

File: types/es6-shim/index.d.ts (hunk 4 above)
Comment (c7hm4r) [inline types/es6-shim/index.d.ts]: That would be exactly like the current ES6 definition in TypeScript https://github.com/Microsoft/TypeScript/blob/master/lib/lib.es6.d.ts#L4826

**Comment by rbuckton** [comment]
Thanks for the contribution!

**Pull request merged**